src/Resources/contao/widgets/PickerWidget.php

```diff
@@ -0,0 +1,240 @@
+	 * @var string
+	 */
+	protected $strTemplate = 'be_widget';
+
+
+	/**
+	 * Load the database object
+	 *
+	 * @param array $arrAttributes
+	 */
+	public function __construct($arrAttributes=null)
+	{
+		$this->import('Database');
+		parent::__construct($arrAttributes);
+	}
+
+
+	/**
+	 * Return an array if the "multiple" attribute is set
+	 *
+	 * @param mixed $varInput
+	 *
+	 * @return mixed
+	 */
+	protected function validator($varInput)
+	{
+		$this->checkValue($varInput);
+
+		if ($this->hasErrors())
+		{
+			return '';
+		}
+
+		// Return the value as usual
+		if ($varInput == '')
+		{
+			if ($this->mandatory)
+			{
+				$this->addError(sprintf($GLOBALS['TL_LANG']['ERR']['mandatory'], $this->strLabel));
+			}
+
+			return '';
+		}
+		elseif (strpos($varInput, ',') === false)
+		{
+			return $this->multiple ? array(intval($varInput)) : intval($varInput);
+		}
+		else
+		{
+			$arrValue = array_map('intval', array_filter(explode(',', $varInput)));
+
+			return $this->multiple ? $arrValue : $arrValue[0];
+		}
+	}
+
+
+	/**
+	 * Check the selected value
+	 *
+	 * @param mixed $varInput
+	 */
+	protected function checkValue($varInput)
+	{
+		if ($varInput == '' || !is_array($this->rootNodes) || !$this->Database->fieldExists('pid', $this->foreignTable))
+		{
+			return;
+		}
+
+		if (strpos($varInput, ',') === false)
+		{
+			$arrIds = array(intval($varInput));
+		}
+		else
+		{
+			$arrIds = array_map('intval', array_filter(explode(',', $varInput)));
+		}
+
+		if (count(array_diff($arrIds, array_merge($this->rootNodes, $this->Database->getChildRecords($this->rootNodes, $this->foreignTable)))) > 0)
+		{
+			$this->addError($GLOBALS['TL_LANG']['ERR']['invalidRecords']);
+		}
+	}
+
+
+	/**
+	 * Generate the record label
+	 *
+	 * @param array $data
+	 *
+	 * @return string
+	 */
+	protected function generateRecordLabel(array $data)
+	{
+		// Label callback
+		if (is_array($this->label_callback))
+		{
+			$callback = $this->label_callback;
+
+			$label = static::importStatic($callback[0])->{$callback[1]}($data, $this->objDca);
+		}
+		elseif (is_callable($callback = $this->label_callback))
+		{
+			$label = $callback($data, $this->objDca);
+		}
+		elseif ($this->labelField)
+		{
+			// Label field
+			$label = $data[$this->labelField];
+		}
+		else
+		{
+			$label = $data['id'];
+		}
+
+		// Add the icon
+		if ($this->icon)
+		{
+			$label = sprintf('%s %s', \Image::getHtml($this->icon, '', 'style="vertical-align:middle;"'), $label);
+		}
+
+		return $label;
+	}
+
+
+	/**
+	 * Generate the widget and return it as string
+	 *
+	 * @return string
+	 */
+	public function generate()
+	{
+		$arrValues = array();
+
+		// Generate the records
+		if (!empty($this->varValue)) // Can be an array
+		{
+			$ids = array_map('intval', (array) $this->varValue);
+			$records = $this->Database->execute("SELECT * FROM {$this->foreignTable} WHERE id IN (" . implode(',', $ids) . ")" . ($this->sortable ? (" ORDER BY " . $this->Database->findInSet('id', $ids)) : ""));
```

Comment:
Generating the query should go to its own protected method so it can be easily overridden in a custom widget.

Reply:
Why would that be necessary? With the label callback everything is possible, so I don't ever see a need to join something? If you don't have actual DB records, I would think one should override the whole method?

Reply:
Imo this should stay as is. What if I want to work on models and not the query? Or non-DB records like @aschempp mentioned?